db/commitlog/commitlog.cc

```diff
@@ -1427,11 +1427,19 @@ future<db::commitlog::segment_manager::sseg_ptr> db::commitlog::segment_manager:
 
     ++_new_counter;
 
-    // don't increase reserve count if we are at max, or we would go over disk limit. 
-    if (_reserve_segments.empty() && (_reserve_segments.max_size() < cfg.max_reserve_segments) && (totals.total_size_on_disk + max_size) <= max_disk_size) {
-        _reserve_segments.set_max_size(_reserve_segments.max_size() + 1);
-        clogger.debug("Increased segment reserve count to {}", _reserve_segments.max_size());
+    if (_reserve_segments.empty()) {        
+        // don't increase reserve count if we are at max, or we would go over disk limit. 
+        if (_reserve_segments.max_size() < cfg.max_reserve_segments && (totals.total_size_on_disk + max_size) <= max_disk_size) {
+            _reserve_segments.set_max_size(_reserve_segments.max_size() + 1);
+            clogger.debug("Increased segment reserve count to {}", _reserve_segments.max_size());
+        }
+        // if we have no reserve and we're above/at limits, make background task a little more eager.
+        if (!_shutdown && totals.total_size_on_disk >= disk_usage_threshold) {
```

Comment:
total_size_on_disk is inaccurate. It is incremented when we put a file to files_to_delete list, but at this point the file is not deleted yet and may be recycled. The following code explicitly re-arms the timer, but it's logic is inaccurate - we may re-arm the timer even though we will end up doing more work on disk, or may re-arm the timer too late, because total_size_on_disk doesn't include the size of reserved segments and the size of segments in the deletion queue (files_to_delete). 
This maybe kind-of works, but there is no guarantee we don't end up deleting the segments, or don't end up waiting for flushes for too long.

I'd start these series from building up accurate metrics for the total size on disk, which includes everything, and designing a mechanism which aligns the speed of flushes with the speed of use of segments. Simply kicking in the flushes when we're at 75% of segment size limit may be too much or too little, depending on the workload.

Reply:
What are you talking about? total_size_on_disk is decremented in one place: (after actual file deletion), and incremented in two: when we allocate disk space (in allocate_segment_ex, we add space written by prealloc, and in segment::flush we add space _added_ (if any) by flush).

Reply:
And it very much includes the size of reserve. If they hold space.

Reply:
total_size_on_disk is decremented in ~segment(), where the segment is added to files_to_delete in function add_file_to_delete(). At this point the file is still taking up space. 
Later on we're looking at the list of files in deciding to maybe not delete the file, and put it into reserved_segments. This happens in delete_segments. At this point we do not increment-back the total size on disk. So the total size on disk doesn't account for reserved segments and it is inaccurate, because it is decremented before a file is deleted *or* added to the list of reserved segments.

Reply:
No, it is not. I don't know which version of the code you are looking at, but it certainly is not this one.

Reply:
I was looking at scylla-dev/master. Somehow it's kept up to date, but it doesn't contain some of the patches. When were your previous patches merged?

Reply:
I have no idea what state scylla-dev is in. These patches are based on scylla master. And are self-contained as such.